spec/cookie_spec.rb

```diff
@@ -0,0 +1,167 @@
+  let(:token_payload) do
+    {
+      data: 'data',
+      exp: Time.now.to_i + 4 * 3600,
+      nbf: Time.now.to_i - 3600,
+      iss: 'https://my.awesome.website/',
+      aud: 'audience',
+      jti: jti_digest,
+      iat: Time.now.to_i,
+      sub: 'subject'
+    }
+  end
+  let(:inner_app) do
+    ->(env) { [200, env, [token_payload.to_json]] }
+  end
+  let(:app) do
+    Rack::JWT::Auth.new(
+      inner_app,
+      decode_args
+    )
+  end
+  let(:token) { Rack::JWT::Token.encode(token_payload, secret, 'HS256') }
+  let(:parsed_body) { JSON.parse(last_response.body, symbolize_names: true) }
+
+  before do
+    allow(Time).to receive(:now).and_return(Time.new(2014, 0o2, 0o6, 16, 12, 0, '-08:00'))
+  end
+
+  after do
+    clear_cookies
+  end
+
+  context 'when the cookie setting is enabled' do
+    let(:decode_args) do
+      {
+        secret: secret,
+        verify: true,
+        options: {
+          algorithm: 'HS256',
+          cookie_name: 'authToken',
+          verify_expiration: true,
+          verify_not_before: true,
+          iss: 'https://my.awesome.website/',
+          verify_iss: true,
+          verify_iat: true,
+          jti: jti_digest,
+          verify_jti: true,
+          aud: 'audience',
+          verify_aud: true,
+          sub: 'subject',
+          verify_sub: true,
+          leeway: 30
+        }
+      }
+    end
+
+    context 'when a valid auth cookie is sent with the request, but no auth header' do
+      it 'decodes the cookie and adds the token to the headers' do
+        set_cookie("authToken=#{token}")
+        get('/')
+
+        aggregate_failures do
+          expect(last_response.status).to eq 200
+          expect(parsed_body).to eq(token_payload)
+          expect(last_response.headers['jwt.header']).to eq('typ' => 'JWT', 'alg' => 'HS256')
+        end
+      end
+    end
+
+    context 'when the cookie and Authorization header are both missing' do
+      it 'decodes the cookie and adds the token to the headers' do
+        get('/')
+
+        aggregate_failures do
+          expect(last_response.status).to eq 401
+          expect(parsed_body).to eq(error: 'Missing token cookie and Authorization header')
+          expect(last_response.headers['jwt.header']).to be_nil
+        end
+      end
+    end
+
+    context 'when a valid Authorization header, but no cookie is sent with the request' do
+      it 'decodes token' do
+        header 'Authorization', "Bearer #{Rack::JWT::Token.encode(token_payload, secret, 'HS256')}"
+        get('/')
+
+        aggregate_failures do
+          expect(last_response.status).to eq 200
+          expect(parsed_body).to eq(token_payload)
+          expect(last_response.headers['jwt.header']).to eq('typ' => 'JWT', 'alg' => 'HS256')
+        end
+      end
+    end
+
+    context 'when a valid Authorization header, and a valid auth cookie are both sent with the request' do
+      it 'decodes token' do
```

Comment:
decodes cookie if both are present, isn't it?

Reply:
Yeah. But both are the same in the spec. I'll adjust to verify the cookie takes precedence